pkg/storage/endpoint/tso.go

```diff
@@ -70,9 +72,31 @@ func (se *StorageEndpoint) LoadTimestamp(prefix string) (time.Time, error) {
 	return maxTSWindow, nil
 }
 
-// SaveTimestamp saves the timestamp to the storage.
-func (se *StorageEndpoint) SaveTimestamp(groupID uint32, ts time.Time) error {
+// SaveTimestamp saves the timestamp to the storage. The leadership is used to check if the current server is leader
+// before saving the timestamp to ensure a strong consistency for persistence of the TSO timestamp window.
+func (se *StorageEndpoint) SaveTimestamp(groupID uint32, ts time.Time, leadership *election.Leadership) error {
+	logFilds := []zap.Field{
+		zap.Uint32("group-id", groupID),
+		zap.Time("ts", ts),
+		zap.String("leader-key", leadership.GetLeaderKey()),
+		zap.String("expected-leader-value", leadership.GetLeaderValue()),
+	}
+	log.Info("saving timestamp to the storage", logFilds...)
+	// Fast path to return error if the leadership has not been granted yet.
+	if len(leadership.GetLeaderValue()) == 0 {
+		return errors.Errorf("%s due to leadership has not been granted yet", errs.NotLeaderErr)
+	}
 	return se.RunInTxn(context.Background(), func(txn kv.Txn) error {
+		// Ensure the current server is leader by reading and comparing the leader value.
+		leaderValue, err := txn.Load(leadership.GetLeaderKey())
+		if err != nil {
+			return err
+		}
+		if expected := leadership.GetLeaderValue(); leaderValue != expected {
```

Comment:
An additional check is needed here. Why can’t we confirm if the leadership matches the etcd persisted data or is empty?

Reply:
We verified whether `leadership.GetLeaderValue()` is empty at line 86. This check is necessary because all TSO persistence operations must occur only after the PD leadership or TSO primary has been successfully retrieved.